internal/cli/cli.go

```diff
@@ -38,83 +42,180 @@ type CommandDefinition struct {
 }
 
 // CommandFactory is a command factory
-type CommandFactory struct {
-	Profile *Profile
-	Config  *Config
+type CommandFactory interface {
+	Build(func() CommandDefinition) *cobra.Command
+	Close()
+	Init()
+	Run(cmd *cobra.Command)
+	SetGlobalFlags(fs *flag.FlagSet)
+}
+
+type commandFactory struct {
+	in      *os.File
+	out     *os.File
+	err     *os.File
+	config  *Config
+	profile *Profile
+	ui      terminal.UI
+	logger  *log.Logger
 }
 
 // Config is the global CLI config
 type Config struct {
-	Command CommandConfig
-	UI      terminal.UIConfig
+	CommandConfig
+	terminal.UIConfig
 }
 
 // CommandConfig holds the global config for a CLI command
 type CommandConfig struct {
 	RealmBaseURL string
 }
 
+// NewCommandFactory creates a new command factory
+func NewCommandFactory() CommandFactory {
+	logger := log.New(os.Stderr, "UTC ERROR ", log.Ltime|log.Lmsgprefix)
+
+	config := new(Config)
+
+	profile, profileErr := NewDefaultProfile()
+	if profileErr != nil {
+		logger.Fatal(profileErr)
+	}
+
+	return &commandFactory{
+		config:  config,
+		profile: profile,
+		logger:  logger,
+	}
+}
+
+func (factory *commandFactory) Init() {
+	if err := factory.profile.Load(); err != nil {
+		factory.logger.Fatal(err)
+	}
+
+	if filepath := factory.config.OutputTarget.Path(); filepath != "" {
+		f, err := os.OpenFile(filepath, os.O_CREATE|os.O_RDWR|os.O_APPEND, 0660)
+		if err != nil {
+			factory.logger.Fatal(fmt.Errorf("failed to open target file: %w", err))
+		}
+		factory.out = f
+	}
+}
+
+func (factory *commandFactory) Close() {
+	if factory.config.OutputTarget.Path() != "" {
+		factory.out.Close()
+	}
+}
+
+type suppressUsageError struct {
+	error
+}
+
+func (factory *commandFactory) Run(cmd *cobra.Command) {
+	if err := cmd.Execute(); err != nil {
+		if _, ok := err.(suppressUsageError); !ok {
+			fmt.Println(cmd.UsageString())
+		}
+
+		if factory.ui != nil {
+			err = factory.ui.Print(terminal.NewErrorLog(err))
+		}
+
+		if err != nil {
+			factory.logger.Fatal(err)
+		}
+		os.Exit(1)
```

Comment:
```suggestion
		if factory.ui != nil {
			err = factory.ui.Print(terminal.NewErrorLog(err))
			if err != nil {
			    factory.logger.Fatal(err)
			}
			os.Exit(1)
		}



	factory.logger.Fatal(err)
```

Reply:
thoughts on this? logger.Fatal seems like it will exit so refactored to os.Exit(1) only when ui isn't available

Reply:
you're right in that `logger.Fatal` exits, and so that's why when the `ui` is available, we'll need to `os.Exit(1)` as well.

full disclosure, this section is linter-driver.  the original version of the code:

```go
if factory.ui == nil {
	factory.logger.Fatal(err)
}
factory.ui.Print(terminal.NewErrorLog(err))
os.Exit(1)
```

the issue was an unchecked error on `factory.ui.Print`

I had no luck playing around with the `.errcheck` file we have (which has successfully ignored linter complaints for unchecked `res.Body.Close()` and `file.Close` errors).  if you've got any tips for this, I'm all ears b/c I think that above code is its simplest form

Reply:
I'm also realizing I could have also solved with this a simple ` // nolint: errcheck` inline with the `Print` call.  If you've no objections there, I'll take that route

Reply:
final resolution, just actually checking the error now (which adds another place for `factory.errLogger.Fatal` to be invoked, but doesn't involve any linting-overrides)